File: phases/12-cryptography-and-security/09-public-key-i-diffie-hellman/code/main.py
"""
Public Key I — Diffie-Hellman
Phase 12 — Cryptography & Security

Working Diffie-Hellman key exchange implementation with
modular exponentiation, keypair generation, shared secret
computation, and MITM attack simulation.
"""

import hashlib
import os


def mod_exp(base: int, exp: int, mod: int) -> int:
    result = 1
    base = base % mod
    while exp > 0:
        if exp & 1:
            result = (result * base) % mod
        base = (base * base) % mod
        exp >>= 1
    return result


def generate_private(p: int) -> int:
    return int.from_bytes(os.urandom(32), "big") % (p - 2) + 2


def generate_keypair(p: int, g: int) -> tuple[int, int]:
    private = generate_private(p)
    public = mod_exp(g, private, p)
    return private, public


def compute_shared_secret(their_pub: int, my_priv: int, p: int) -> int:
    return mod_exp(their_pub, my_priv, p)


def derive_key(shared_secret: int) -> bytes:
    return hashlib.sha256(str(shared_secret).encode()).digest()[:16]


def simulate_dh(p: int, g: int) -> None:
    print("=== Diffie-Hellman Key Exchange ===")
    print(f"Public parameters: p = {p}, g = {g}")

    alice_priv, alice_pub = generate_keypair(p, g)
    bob_priv, bob_pub = generate_keypair(p, g)

    print(f"\nAlice's private key: {alice_priv}")
    print(f"Alice's public key:  {alice_pub}")
    print(f"Bob's private key:   {bob_priv}")
    print(f"Bob's public key:    {bob_pub}")

    alice_shared = compute_shared_secret(bob_pub, alice_priv, p)
    bob_shared = compute_shared_secret(alice_pub, bob_priv, p)

    print(f"\nAlice's shared secret: {alice_shared}")
    print(f"Bob's shared secret:   {bob_shared}")
    print(f"Shared secrets match:  {alice_shared == bob_shared}")

    alice_key = derive_key(alice_shared)
    bob_key = derive_key(bob_shared)
    print(f"Derived AES-128 key:   {alice_key.hex()}")
    print(f"Bob's key matches:     {alice_key == bob_key}\n")


def simulate_mitm(p: int, g: int, message: str) -> None:
    print("=== Man-in-the-Middle Attack ===")
    print(f"Original message: \"{message}\"\n")

    alice_priv, alice_pub = generate_keypair(p, g)
    bob_priv, bob_pub = generate_keypair(p, g)
    mallory_priv, mallory_pub = generate_keypair(p, g)

    print(f"Alice sends public key:    {alice_pub}")
    print(f"Mallory intercepts, sends: {mallory_pub}")
    print(f"Bob sends public key:      {bob_pub}")
    print(f"Mallory intercepts, sends: {mallory_pub}\n")

    alice_mallory_shared = compute_shared_secret(mallory_pub, alice_priv, p)
    mallory_alice_shared = compute_shared_secret(alice_pub, mallory_priv, p)
    bob_mallory_shared = compute_shared_secret(mallory_pub, bob_priv, p)
    mallory_bob_shared = compute_shared_secret(bob_pub, mallory_priv, p)

    assert alice_mallory_shared == mallory_alice_shared
    assert bob_mallory_shared == mallory_bob_shared

    alice_key = derive_key(alice_mallory_shared)
    mallory_alice_key = derive_key(mallory_alice_shared)
    bob_key = derive_key(bob_mallory_shared)
    mallory_bob_key = derive_key(mallory_bob_shared)

    msg_bytes = message.encode()
    key_len = min(len(alice_key), len(msg_bytes))
    cipher_alice = bytes(a ^ b for a, b in zip(msg_bytes[:key_len], alice_key[:key_len]))
    intercepted = bytes(a ^ b for a, b in zip(cipher_alice, mallory_alice_key[:key_len]))
    cipher_to_bob = bytes(a ^ b for a, b in zip(intercepted, bob_key[:key_len]))

    print(f"Alice encrypts:         {cipher_alice.hex()}")
    print(f"Mallory decrypts, sees: \"{intercepted.decode()}\"")
    print(f"Mallory re-encrypts:    {cipher_to_bob.hex()}")
    print("Neither Alice nor Bob detects the interception!\n")


def main() -> None:
    # RFC 7919 2048-bit safe prime (MODP group 14)
    p_hex = (
        "FFFFFFFFFFFFFFFFC90FDAA22168C234C4C6628B80DC1CD1"
        "29024E088A67CC74020BBEA63B139B22514A08798E3404DD"
        "EF9519B3CD3A431B302B0A6DF25F14374FE1356D6D51C245"
        "E485B576625E7EC6F44C42E9A637ED6B0BFF5CB6F406B7ED"
        "EE386BFB5A899FA5AE9F24117C4B1FE649286651ECE45B3D"
        "C2007CB8A163BF0598DA48361C55D39A69163FA8FD24CF5F"
        "83655D23DCA3AD961C62F356208552BB9ED529077096966D"
        "670C354E4ABC9804F1746C08CA18217C32905E462E36CE3B"
        "E39E772C180E86039B2783A2EC07A28FB5C55DF06F4C52C9"
        "DE2BCBF6955817183995497CEA956AE515D2261898FA0510"
        "15728E5A8AACAA68FFFFFFFFFFFFFFFF"
    )
    p = int(p_hex, 16)
    g = 2

    simulate_dh(p, g)
    simulate_mitm(p, g, "Transfer $1000000 to Mallory")


if __name__ == "__main__":
    main()

File: phases/12-cryptography-and-security/09-public-key-i-diffie-hellman/code/main.rs
// Public Key I — Diffie-Hellman
// Phase 12 — Cryptography & Security
//
// Working Diffie-Hellman implementation with modular exponentiation,
// keypair generation, shared secret computation, and performance
// benchmarking across multiple parameter sets.

use std::time::Instant;

/// Modular exponentiation: (base^exp) % modulus
/// Uses square-and-multiply with u128 intermediates to avoid overflow.
fn mod_pow(mut base: u64, mut exp: u64, modulus: u64) -> u64 {
    let mut result = 1u64;
    base %= modulus;
    while exp > 0 {
        if exp & 1 == 1 {
            result = (result as u128 * base as u128 % modulus as u128) as u64;
        }
        base = (base as u128 * base as u128 % modulus as u128) as u64;
        exp >>= 1;
    }
    result
}

/// Generate a random private key in [2, p-2]
fn generate_private(p: u64) -> u64 {
    use std::time::{SystemTime, UNIX_EPOCH};
    let seed = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap()
        .as_nanos();
    // Simple LCG for demo purposes (not cryptographically secure RNG)
    let mut state = seed as u64;
    state = state.wrapping_mul(6364136223846793005).wrapping_add(1442695040888963407);
    let range = p.saturating_sub(3);
    2 + (state % range)
}

/// Generate a DH keypair: (private, public)
fn generate_keypair(p: u64, g: u64) -> (u64, u64) {
    let private = generate_private(p);
    let public = mod_pow(g, private, p);
    (private, public)
}

/// Compute the shared secret: their_pub^my_priv mod p
fn compute_shared_secret(their_pub: u64, my_priv: u64, p: u64) -> u64 {
    mod_pow(their_pub, my_priv, p)
}

/// Demonstrate DH with a given parameter set (p, g name, p hex suffix)
fn demonstrate(p: u64, g: u64, label: &str) {
    println!("--- Diffie-Hellman with {label} ---");
    println!("p = {p}, g = {g}");

    let start = Instant::now();
    let (alice_priv, alice_pub) = generate_keypair(p, g);
    let gen_time = start.elapsed();

    let (bob_priv, bob_pub) = generate_keypair(p, g);

    let alice_shared = compute_shared_secret(bob_pub, alice_priv, p);
    let bob_shared = compute_shared_secret(alice_pub, bob_priv, p);

    println!("Alice pub: {alice_pub}, Bob pub: {bob_pub}");
    println!("Shared secret: {alice_shared}");
    println!("Match: {}\n", alice_shared == bob_shared);
    println!("Key generation took: {:?}\n", gen_time);
}

/// Simulate a MITM attack (demonstrates the vulnerability)
fn demonstrate_mitm(p: u64, g: u64) {
    println!("--- MITM Attack Simulation ---");

    let (alice_priv, alice_pub) = generate_keypair(p, g);
    let (bob_priv, bob_pub) = generate_keypair(p, g);
    let (mallory_priv, mallory_pub) = generate_keypair(p, g);

    // Alice thinks she's talking to Bob, but Mallory intercepted
    let alice_shared = compute_shared_secret(mallory_pub, alice_priv, p);
    let mallory_with_alice = compute_shared_secret(alice_pub, mallory_priv, p);

    // Bob thinks he's talking to Alice, but Mallory intercepted
    let bob_shared = compute_shared_secret(mallory_pub, bob_priv, p);
    let mallory_with_bob = compute_shared_secret(bob_pub, mallory_priv, p);

    assert_eq!(alice_shared, mallory_with_alice);
    assert_eq!(bob_shared, mallory_with_bob);

    println!("Alice-Mallory shared secret: {alice_shared}");
    println!("Bob-Mallory shared secret:   {bob_shared}");
    println!("Alice and Bob do NOT share a secret directly.");
    println!("Mallory can decrypt, read, and re-encrypt all traffic.\n");
}

fn main() {
    // Small parameters for 64-bit demonstration
    // These are NOT secure — real DH uses 2048+ bit primes.
    demonstrate(9973, 2, "10-bit prime (insecure, for demo)");
    demonstrate(99991, 5, "17-bit prime (insecure, for demo)");

    demonstrate_mitm(9973, 2);

    // Benchmark mod_pow with various exponents
    println!("--- Exponentiation Benchmark ---");
    let p = 99991u64;
    let g = 2u64;
    for exp_bits in [8, 16, 24, 32] {
        let exp = (1u64 << exp_bits) - 1;
        let start = Instant::now();
        let result = mod_pow(g, exp, p);
        let elapsed = start.elapsed();
        println!("mod_pow(2, 2^{exp_bits}-1, 99991) = {result:>5}  [{elapsed:?}]");
    }
}
